test/python/synthesis/test_linear_synthesis.py

```diff
@@ -139,6 +140,29 @@ def test_synth_lnn_kms(self, num_qubits):
                 dist = abs(q0 - q1)
                 self.assertEqual(dist, 1)
 
+    def test_pmh_section_sizes(self):
+        """Test the PMH algorithm for different section sizes.
+
+        Regression test of Qiskit/qiskit#12166.
+        """
+        n = 5
+        qc = QuantumCircuit(n)
+        for i in range(n - 1):
+            qc.cx(i, i + 1)
+        expected = Operator(qc)
+
+        mat = LinearFunction(qc).linear
+        for section_size in range(1, n + 1):
+            synthesized = synth_cnot_count_full_pmh(mat, section_size=section_size)
+            with self.subTest(section_size=section_size):
+                self.assertEqual(expected, Operator(synthesized))
```

Comment:
instead of using `Operator`, you can use: `mat_synth = LinearFunction(qc).linear` and check that it's identical to the original `mat`. The matrix is `nxn` and not `(2^n)x(2^n)`, so you can also check larger values of `n`.

Reply:
I just checked the smallest value for which we saw failures 🙂 I'll increase a bit.